### ISSUE TITLE
Proper clarification on what content Northstar.Custom is supposed to be and not to be.

### ISSUE THREAD

**Issue by EM4Volts** [R2Northstar/NorthstarMods#686]
Over the course of having the pleassure to be in this community i have heard many contributors talk about what Northstar Custom is supposed to be containing, and alot of people seem to think differently of it.

some say its just simple custom stuff like simple gamemodes while others somehow think its a repository on basically pring everything people want ingame, from new weapons to maps to whatever just because what else.

I can, however, not find a single place to find an actual statement on this.
And since the discussion of "wait for autodl" came up again it like to ask for a proper explainer on this as people start thinking that we cant wait for that and instead just should pr everything for the sake of autodl not existing.

**Comment by GeckoEidechse** [comment]
IMO, assuming we can get mod-auto-download out in a reasonably near future, I'd say `Northstar.Custom` is anything that would be considered like a "core Northstar feature" that is not part of the base game.

So things like _Frontier War_ (cause it was never released in base), _Fastball_ (just cause it's been in Northstar from the beginning and is kinda considered one of its most iconic modes even if it gets barely played now), and _Frontier Defense_ (cause we need to rebuild it from scratch basically).

One thing to consider is that due to our shitty release pipeline (1/3 of users manually updating, some mod-managers not deleting old files), we essentially atm cannot remove any added source files again, meaning that any new source file added also adds technical debt that cannot be removed. This is one of the primary reasons to push forward auto-mod download (and improvements to release pipeline).

Also, IIRC without `Northstar.Custom` Northstar should have vanilla compatibility. (which also means that technically `Northstar.CustomServers` should've been `Northstar.Servers`  but that's a whole other discussion).

**Comment by ASpoonPlaysGames** [comment]
> assuming we can get mod-auto-download out in a reasonably near future, I'd say `Northstar.Custom` is anything that would be considered like a "core Northstar feature" that is not part of the base game

For the most part, yeah I agree. However by moving new content to separate mods, we would lose most control over quality, versioning, etc. 

IMO when autodownload becomes a thing, Northstar.Custom should be more like a collection of "official" mods that provide a new user with a solid amount of new modded features. It should provide an "authentic" northstar experience that the user can modify further through other mods.

> So things like _Frontier War_ (cause it was never released in base), _Fastball_ (just cause it's been in Northstar from the beginning and is kinda considered one of its most iconic modes even if it gets barely played now), and _Frontier Defense_ (cause we need to rebuild it from scratch basically).

Frontier defense should be in `Northstar.CustomServers` as it is a recreation of vanilla behaviour, not custom content.

> Also, IIRC without `Northstar.Custom` Northstar should have vanilla compatibility

For a vanilla client connecting to a Northstar server, yes. For a northstar client connecting to a vanilla servers, no, `Northstar.Client` breaks that compatibility (as well as `-norestrictservercommands`)

> (which also means that technically `Northstar.CustomServers` should've been `Northstar.Servers` but that's a whole other discussion).

I think you misunderstand the point of `Northstar.CustomServers` tbh. It's main function is to let people host servers, but also to provide a framework to help people created modded (custom) servers. 

The mod's description I think is pretty clear:
> Attempts to recreate the behaviour of vanilla Titanfall 2 servers, as well as changing some scripts to allow better support for mods

**Comment by GeckoEidechse** [comment]
> Frontier defense should be in `Northstar.CustomServers` as it is a recreation of vanilla behaviour, not custom content.

Whoops, my mistake. Internet is barely functioning rn so I couldn't check where the FD branch PRs into ^^"

&nbsp;

> I think you misunderstand the point of `Northstar.CustomServers` tbh. It's main function is to let people host servers, but also to provide a framework to help people created modded (custom) servers.

Yeah, that was basically the point of the whole discussion back in the day to sorta split up the content needed for hosting servers and the part that adds non-vanilla functionality that is server-side only.

&nbsp;

> > assuming we can get mod-auto-download out in a reasonably near future, I'd say `Northstar.Custom` is anything that would be considered like a "core Northstar feature" that is not part of the base game
> 
> For the most part, yeah I agree. However by moving new content to separate mods, we would lose most control over quality, versioning, etc.

Yes, basically

> IMO when autodownload becomes a thing, Northstar.Custom should be more like a collection of "official" mods that provide a new user with a solid amount of new modded features. It should provide an "authentic" northstar experience that the user can modify further through other mods.

summarises it best IMO.

**Comment by Zanieon** [comment]
Mentioning my perspective on this: I pratically share the same vision about these folders with Spoon, the only difference is that `Northstar.CustomServers` shall be a framework for modded server within the boundaries of vanilla featureset, even more so because we have a compile check here on github that ensures that folder works without the need of `Northstar.Custom`, which have basically all Northstar own framework for deeper modding (i.e sh_message_utils or SafeIO scripts).

Might sound hypocrite hearing this from me because my own FD branch doesn't do that, but that's only because i delayed to understand properly what is the true role each of those folders plays, but now what i will eventually do is move all experimentals i did for Frontier Defense into Northstar.Custom while leaving the gamemode_fd of Northstar.CustomServer strictly with only vanilla features and essential fixes.

**Comment by EladNLG** [comment]
I do not agree - if we just leave northstar to die and rely on "third-party" mods to keep it alive, that's an easy death sentence.

As of right now, autodownload hasn't been touched in over 3 months, outside of discord discussions. "Waiting for autodownload" cannot be used to deny any content updates that aren't related to vanilla - because if that was the original intent of northstar (it isn't), Gun Game, Sticks and Stones, the Peacekraber, the Arc Cannon, the Double Barrel (and no, they are not vanilla, as the weapon files did not exist for them in Titanfall 2), One in the Chamber, Hide and Seek, and many more features wouldn't have been merged.

Northstar shouldn't be just these features forever - because as of right now? Working on mods that are both server & client is heavily discouraged, and until auto-download is _released_, not "being worked on", **_released_**, no one will want to make these mods because no one will play with them and it'll be a huge waste of time for the modder.

I wasted like 3 months on Economy and people told me to "wait for autodownload" over a year ago. Pretending that it is a good reason to deny merges of more content is utterly ridiculous.

**Comment by EnderBoy9217** [comment]
I'm between Elad and Spoon's statements.

On one hand we should be added content to base Northstar, as we already have broken the trend of vanilla, we will need reasons for people to come back to Northstar and should populate new releases with new, fun, and improved features and mechanics.

But to Spoon's point, anything released under Northstar should be "Official" and should be held to a very high quality standard and shouldn't be unintuitive or confusing in any way.

And while I get the statements that "Northstar shouldn't be adding new stuff/keep it vanilla" and "mod autodownload Soon(TM) please wait for that", I know that many users would love seeing things such as new gamemodes/riffs/weapons/other custom content.

Of course, I don't believe would should force any changes or modifications to vanilla stuff (E.G. Rebalance mods/skins/other overrides should never be in Northstar)

**Comment by GeckoEidechse** [comment]
> And while I get the statements that "Northstar shouldn't be adding new stuff/keep it vanilla" [...]

This was never said. Rather my main point is to keep in mind whether adding certain content is worth accumulating the technical debt it costs to maintain it.
Again we currently **cannot** remove content we add which means if we just mess up **once** it's gonna cost us a lot of tickets and announcements to remedy.

> [...] "mod autodownload Soon(TM) please wait for that", I know that many users would love seeing things such as new gamemodes/riffs/weapons/other custom content.

Until we have mod-auto-download, we basically have to guess whether a certain gamemode that we could add would turn out to become popular (Raise your hand if you still play _Hide and Seek_ or _One in the Chamber_ in 2023). The last thing we wanna do is add new modes just for them to end up dead in water (RIP _Frontier War_).

With auto-mod-download, new custom gamemodes become trivial to install and join, and if we notice a particular gamemode becoming popular we can then add it to base Northstar. (Sorta how _Mod Settings_ has since been added to Northstar).

&nbsp;

> As of right now, autodownload hasn't been touched in over 3 months, outside of discord discussions.

Keep in mind that everyone is working on this project on their free time. Sometimes life gets in the way. Sometimes, if you want to see something completed that doesn't seem to make progress you have to step in yourself.

> I do not agree - if we just leave northstar to die and rely on "third-party" mods to keep it alive, that's an easy death sentence.

Northstar will also "die" if we keep accumulating technical debt to the point where the project becomes unmaintainable.

We already deal with cases where the person that wrote the original lines of code is no longer available and you have to spend additional time to clean it up.

&nbsp;

Ultimately, the best approach would be to go a similar route as r5r does where there's some base SDK that exposes various game logic that then a set of "official" and 3rd party mods can built on top of. However that would also require lots more of engine reversing and outside of Blue digging through RUI, I'm not aware of anyone doing that atm.

**Comment by F1F7Y** [comment]
Replying to @GeckoEidechse, not quoting as I'm on mobile

I fully agree on being strict with custom content. That being said the technical debt argument goes both ways and pushing for Auto-Download now without refactoring core systems will just make our life harder in the future.

I have a branch where I'm taking my own spin on Northstar ( called `primedev` on my fork ). My goal is to move towards source engine file structure and removing autohook and offset_structs ( which requires reversing ) while still being fully northstar compatible. Ig the main goal is to have an example I can point to as many discord discussions are just he said she said with no ground to stand on.
